Tools/parametric_model/src/tools/dataframe_tools.py

```diff
@@ -32,35 +32,102 @@
+
+        if control_threshold is None:
+            control_threshold = ACTUATOR_CONTROLS_THRESHOLD
+
+        actuator_ulogs = config_dict['data']['required_ulog_topics']['actuator_outputs']['ulog_name']
+        actuator_ulogs = [x for x in actuator_ulogs if x != 'timestamp']
+
+        act_df_crp = act_df.copy()
+
+        for actuator_topic in actuator_ulogs:
+            act_df_crp = act_df_crp[act_df_crp[actuator_topic] > pwm_threshold]
+
+        t_start = act_df_crp.iloc[1, 0]
+        t_end = act_df_crp.iloc[(act_df_crp.shape[0]-1), 0]
+        flight_time = {"t_start": t_start, "t_end": t_end}
+
+        # Alternative option: use the entire log for model identification (might be problematic with ground phases of the log)
+        # t_start = act_df.iloc[0, 0]
+        # t_end = act_df.iloc[(act_df.shape[0]-1), 0]
+        # flight_time = {"t_start": t_start, "t_end": t_end}
+
+    else:
+        flight_time = []
+        use_with_thrust_only = config_dict.get('use_with_thrust_only', False)
+        use_zero_thrust_only = config_dict.get('use_zero_thrust_only', False)
+
+        zero_crossings = np.where(
+            np.diff(np.sign(act_df['aux1'] + (act_df['aux1'] == 0))))[0]
+
+        assert len(zero_crossings ==
+                   0), "No aux1 activations detected as required for the ramp flight mode"
+        assert len(
+            zero_crossings) % 2 == 0, "Aux1 active ranges are not all closed"
+
+        num_of_ramps = len(zero_crossings) // 2
+        print('\nDetected {0} ramp inputs in total'.format(num_of_ramps))
+
+        throttle_list = []
+        num_thrust_ramps = 0
+        num_pitch_ramps = 0
+
+        for i in range(num_of_ramps):
+            new_flight_time = {"t_start": act_df['timestamp'][zero_crossings[2 * i]],
+                               "t_end": act_df['timestamp'][zero_crossings[2 * i + 1]]}
+
+            ramp_throttle = thrust_df[np.logical_and((thrust_df['timestamp'] > new_flight_time['t_start']),
+                                      (thrust_df['timestamp'] < new_flight_time['t_end']))]
+            throttle_list.append(ramp_throttle)
+
+            num_non_zero = ramp_throttle[ramp_throttle['xyz[0]']
+                                         > 0.0]['xyz[0]'].shape[0]
+            if num_non_zero > 0:
+                num_thrust_ramps += 1
+                if use_zero_thrust_only:
+                    continue
+            else:
+                num_pitch_ramps += 1
+                if use_with_thrust_only:
+                    continue
+
+            flight_time.append(new_flight_time)
 
-    if pwm_threshold is None:
-        pwm_threshold = PWM_THRESHOLD
+        print('\nProcesed {0} thrust ramps and {1} pitch ramps'.format(
+            num_thrust_ramps, num_pitch_ramps))
 
-    if control_threshold is None:
-        control_threshold = ACTUATOR_CONTROLS_THRESHOLD
-    act_df_crp = act_df[act_df.iloc[:, 2] > pwm_threshold]
-    act_df_crp = act_df[act_df.iloc[:, 4] > pwm_threshold]
```

Comment:
If I understand it correctly, the line before this one won't have an effect as the result is overwritten anyways? The new implementation that checks if all actuator outputs are above the pwm threshold (if no selection parameter is used) is probably too restrictive on the other way to detect if the aircraft is airborne... What was the idea behind this flight detection?

Reply:
The flight time computation has been completely reworked now and is based on the 'landed' prop of the vehicle status topic. This should be more reliable and generally applicable to many different types of aircraft.